### ISSUE TITLE
Python version support

### ISSUE THREAD

**Issue by lwasser** [stravalib/stravalib#254]
Hey y'all - i think we've informally decided to drop python 2.x which is a breaking change but i think really necessary. Also when building i think we should drop 3.6 which i @jsamoocha did. i just want to ensure that we are all on the same page as i'm modernizing the packaging infrastructure i also am dropping support for Python =< 3.6

[Here is the pr opened in my repo so i could play with CI](https://github.com/lwasser/stravalib/pull/7/files) - i'm not fully sure about poetry yet but the other changes i think are necessary to update our package's structure. are we all on the same page:

Drop support for. 
* python 2.x
* drop 3.6 or lower
And then related to that the current version is

version = "0.10.4"

Given we have some breaking changes would we bump to 0.11 here? we aren't quite at 1.0 given all of hte refactoring but this isn't exactly a minor update we are all doing. 

Many thanks for input @jsamoocha @hozn @yihong0618 !! Please respond that you are in support of (or not) dropping 2.x and 3.6 and lower. and then let me know what version we should bump to for this initial release.

**Comment by jsamoocha** [comment]
I'm not sure about what problem is solved by poetry in this case. Will it force developers to also use poetry if they want to work with stravalib? With a plain `requirements.txt` file, everybody can at least use their own preferred virtual env + dependency management system.

Besides that, python >=3.7 and version 0.11 look great to me.

Just out of curiosity (I'm not an expert on versioning/releases), is there any (automated) relationship between: 

- the version mentioned in pyproject.toml,
- a tag/release pushed to this repo,
- the actual shipping of a release to pypi?

**Comment by lwasser** [comment]
@jsamoocha so happy you responded here as i'm really super torn on what tools we should use. 

So it sounds like we all support python >3.7 as being our python base. perfect.

# TLDR 

* Poetry is ideal for dependency management. Just because we use it to manage dependency and build our wheel doesn't mean that all maintainers / contributors have to use it. We could keep a requirements file too - 
* Poetry might make it easy for l ocal builds of docs and such `poetry build docs` vs creating a make or nox file (i prefer nox).
* python semantic release is for versioning - and requires semantic commit messages. We could enforce those when merging pr's but it could be a bit more effort to think about commit names and be consistent. To automate we'd setup a workflow_dipatch item in ci and would click a button to bump and push to pypi. What is nice here is the tool manages what version so we aren't typing in a version when creating a tag. 
* setuptools scm is for versioning too. It relies on a tag to trigger a build. in this case when we create a new tag the build / pypi workflow would trigger.

# LONG VERSION: :)

## Poetry - package structure and dependency management and build 

The biggest thing that poetry does for us IMO is [ lock down dependencies](https://github.com/lwasser/stravalib/pull/7/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R9). Essentially by running poetry add --group -dev you can add a dev dependency or a -doc dependency or a package dependency and it will resolve and pin the best versions for each to ensure the package builds across all of our specified python supported versions. 

Poetry also has a poetry version to bump locally and poetry build to build the wheel, etc. but honestly those actions are also easily done with other tools like twine and such to build so i'm less concerned about that. Poetry does support us adding diff virtual envt managers for instance if we wanted to use conda instead it's easy to tell it to use conda. But you are correct if someone wanted to use something different from what we (maintainers) use, they'd have to modify the poetry config and i haven't played with that yet. OR honestly they could just not use poetry for their environment it hink.  IMO we'd select poetry for easy of managing of dependencies. 

NOTE: We can always keep a requirements.txt file for people as well.

Ok bare with me - two more tools - i know this is a lot of info. i've just been researching to pick something that we all like and is easy to maintain. 

# VERSION management 

## setuptools-scm 

A lot of people are using setuptools scm to manage  versions. setuptools scm is nice because it's tag based. so when we (manually) create a new tag, and push it triggers setuptools to do it's thing. i think setup tools also adds a hash associated with a bump version but i plan to play with this a bit more today / tomorrow. This hash is really nice if you are using a tool and need to fix environment conflicts.

### Setuptools SCM and automation
 setuptools scm doesn't have great documentation [but here is a build that uses it. ](https://github.com/fatiando/verde/blob/main/.github/workflows/pypi.yml). i really like how leo (maintainer there) builds his packages.

With setuptools scm we'd setup an automated build like the one linked to above. And the release would look like this:
* one of us updates the changelog,  creates a new tag and i think a release - i need to look to see if we'd do a tag and release or just a tag. its not too hard to trigger an auto release but we probably want to edit the text... we can talk thru this separately.
* The release then triggers a build. We'd have a test built to pypi test to ensure all is good. and if all is good it publishes on the regular pypi. The entire build can thus be triggered from Github.com!  

## Python Semantic release and automation

[Python semantic version](https://github.com/python-semantic-release/python-semantic-release) is another tool i've been playing with. 
We could still implement the automated workflow that i described above however it would be triggered a bit differently. with python semantic release (PSR). So PSR requires that we use a semantic commit name workflow. so things like 

Fix(docs): message here would trigger a bump patch 

Essentially it enforces semantic commit messages and determines what version to bump to based upon that. We could always manage the commit text in pr's when we squash and merge. We also can setup a bot to check for semantic commit messages on prs (or a precommit hook for local checks). 

BENEFITS:
* PSR also helps to automate the process of creating a change log. It updates it for us on build. With those semantic commit messages it becomes easier to track change log automatically. 
* PSR enforces semantic commit messages which are more intuitive to look at in the git log 
* PSR makes decisions about bump version for us based upon the type of change we make to the code

My concern with PSR is as nice as semantic messages are, i worry it would be easy to get a word wrong and thus cause an additional layer of effort for maintainers. 

BTW 
* we can use any of the two version management tools above w/ poetry
* We can automate the build and push to pypi with any combination of tools as well. 

## Changelog
In the above tools, PSR automates the changelog. Whereas with setuptools SMT we are manually updating the changelog. 
I'm kind of leaning towards a manual update of the changelog on prs as its not really hard to do we just include in our pr check list (create. template)

*************

I hope that isn't' TMI!! 😆 I would LOVE some input on what type of build workflow sounds easiest to you all to maintain and use.

**Comment by jsamoocha** [comment]
Wow, thanks for the thorough explainer!

> My concern with PSR is as nice as semantic messages are, i worry it would be easy to get a word wrong and thus cause an additional layer of effort for maintainers.

I think I share this concern. I'm not sure that all contributors to this package (including myself) have the required discipline (yet) to work with semantic commit messages. Maybe (but please correct me if I'm wrong) ad-hoc / tag based releases allow a bit more flexibility?

**Comment by lwasser** [comment]
Ok great. SO i will close this. I"ve moved away from poetry and semantic release for the time being. We might reconsider something like poetry or flit in the future... but for now i think we are ok with just using vanilla `build`. thank you for the input @jsamoocha !! closing!!
